tftp_test.go

```diff
@@ -0,0 +1,211 @@
+package ipxe
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pin/tftp"
+	"github.com/tinkerbell/boots-ipxe/binary"
+	"go.opentelemetry.io/otel/trace"
+	"inet.af/netaddr"
+)
+
+type fakeReaderFrom struct {
+	addr    net.UDPAddr
+	content []byte
+	err     error
+}
+
+func (f *fakeReaderFrom) ReadFrom(r io.Reader) (n int64, err error) {
+	if f.err != nil {
+		return 0, f.err
+	}
+	nInt, err := r.Read(f.content)
+	return int64(nInt), err
+}
+
+func (f *fakeReaderFrom) SetSize(_ int64) {}
+
+func (f *fakeReaderFrom) RemoteAddr() net.UDPAddr {
+	return f.addr
+}
+
+func (f *fakeReaderFrom) WriteTo(_ io.Writer) (n int64, err error) {
+	return 0, nil
+}
+
+func TestListenAndServeTFTP(t *testing.T) {
+	ht := &HandleTFTP{Log: logr.Discard()}
+	srv := tftp.NewServer(ht.ReadHandler, ht.WriteHandler)
+	type args struct {
+		ctx  context.Context
+		addr netaddr.IPPort
+		h    *tftp.Server
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr interface{}
+	}{
+		{
+			name: "fail",
+			args: args{
+				ctx:  context.Background(),
+				addr: netaddr.IPPortFrom(netaddr.IPv4(127, 0, 0, 1), 80),
+				h:    nil,
+			},
+			wantErr: &net.OpError{},
+		},
+		{
+			name: "success",
+			args: args{
+				ctx:  func() context.Context { c, cn := context.WithCancel(context.Background()); defer cn(); return c }(),
```

Comment:
isn't this racy? The context is being cancelled as you're returning it to the caller

Reply:
ah, this is left over from some testing. shouldn't be here. good catch. will remove it.

Reply:
:+1: